kf_lib_data_ingest/target_apis/kids_first.py

```diff
@@ -23,497 +22,690 @@ def indexd_hashes(dictstr):
     }
 
 
-target_concepts = {
-    "investigator": {
-        "standard_concept": CONCEPT.INVESTIGATOR,
-        "properties": {
-            target_service_entity_id: CONCEPT.INVESTIGATOR.TARGET_SERVICE_ID,
-            "external_id": CONCEPT.INVESTIGATOR.UNIQUE_KEY,
-            "name": CONCEPT.INVESTIGATOR.NAME,
-            "institution": CONCEPT.INVESTIGATOR.INSTITUTION,
-            "visible": CONCEPT.INVESTIGATOR.VISIBLE,
-        },
-        "endpoint": "/investigators",
-    },
-    "study": {
-        "standard_concept": CONCEPT.STUDY,
-        "links": [
+def join(*args):
+    return ".".join([str(a) for a in args])
+
+
+def get_target_id(entity_class, row, target_id_lookup_func):
+    tic = row.get(entity_class.target_id_concept)
+
+    if tic and (tic != constants.COMMON.NOT_REPORTED):
+        return tic
+    else:
+        return target_id_lookup_func(
+            entity_class.__name__, entity_class.build_key(row)
+        )
+
+
+def without_nulls(body):
+    # Remove elements with null values
+    return {k: v for k, v in body.items() if v is not None}
+
+
+class Investigator:
```

Comment:
I explored this but decided against it. The only reason to do that would be to check and enforce in the loading code itself that every class has the base as its super, but there's not much of a point to doing that since it would fail on calling the functions anyway if they're not implemented.

Reply:
Hmm I guess I'm curious to see what the error messages + stack trace will look like with what you have now (I'll try it out). I guess as long as they are informative to the developer and the developer understands what is required of their entity classes, then ok. 

Also RE: documenting required methods, as long as we document the target API entity class's requirements (not just in the docstrings, but in our lib docs) well, then I think I'll feel better about it. 

I keep thinking about how in the near future more than just our team may use this library (for FHIR ingest) so I want to make sure error messaging, documentation is very clear. I figured that abcs are the standard way define and document the interface that developers need to make their subclasses conform to. Anything that differs from that might confuse people? Idk that was my original thought, but I'll try out the lib to see what the dev experience is like